File: Desktop_GUI/CognitoHandler.py
import os
import time

import boto3
from botocore.exceptions import ClientError

# Configura tus datos de Cognito
USER_POOL_ID = "us-east-1_WRwcwjyj4"
CLIENT_ID = "2nr4s8pcc36ppp5o212mpcmfmj"

os.environ['AWS_DEFAULT_REGION'] = 'us-east-1'

class CognitoHandler:
    def __init__(self, username: str, password: str):
        self.client = boto3.client('cognito-idp')

        self.username = username
        self.password = password

        self.tokens = None

    def login(self):
        try:
            response = self.client.initiate_auth(
                AuthFlow="USER_PASSWORD_AUTH",
                AuthParameters={
                    "USERNAME": self.username,
                    "PASSWORD": self.password
                },
                ClientId=CLIENT_ID
            )

            self.tokens = response['AuthenticationResult']
            return self.tokens

        except ClientError as e:
            raise Exception(e.response['Error']['Message'])

    def logout(self):
        if self.tokens and self.tokens.get('AccessToken', None):
            self.client.global_sign_out(
                AccessToken=self.tokens['AccessToken'],
            )

    def refresh_token(self):
        client = boto3.client("cognito-idp")
        response = client.initiate_auth(
            AuthFlow="REFRESH_TOKEN_AUTH",
            AuthParameters={
                "REFRESH_TOKEN": self.tokens['RefreshToken']
            },
            ClientId=CLIENT_ID
        )

        self.tokens = response['AuthenticationResult']

    def get_access_token(self) -> str | None:
        if self.tokens and self.tokens.get('AccessToken', None):
            return self.tokens['AccessToken']
        return None

    def get_valid_access_token(self):
        if not self.tokens:
            raise Exception("⚠️ No hay sesión activa")

        if time.time() > self.tokens["expires_at"] - 60:  # Renovar si faltan <60s
            print("🔁 Token expirado o por expirar. Renovating...")
            new_tokens = self.refresh_token(self.tokens["RefreshToken"])
            return new_tokens["AccessToken"]
        else:
            return self.tokens["AccessToken"]

File: electron/cognito-handler.js
const { CognitoIdentityProviderClient, InitiateAuthCommand, GetUserCommand, GlobalSignOutCommand } = require('@aws-sdk/client-cognito-identity-provider');

// Configuración de Cognito (mismos valores que Python)
const USER_POOL_ID = "us-east-1_WRwcwjyj4";
const CLIENT_ID = "2nr4s8pcc36ppp5o212mpcmfmj";
const AWS_REGION = "us-east-1";

class CognitoHandler {
    constructor() {
        this.client = new CognitoIdentityProviderClient({ 
            region: AWS_REGION 
        });
        this.tokens = null;
        this.username = null;
    }

    async login(username, password) {
        try {
            const command = new InitiateAuthCommand({
                AuthFlow: "USER_PASSWORD_AUTH",
                AuthParameters: {
                    USERNAME: username,
                    PASSWORD: password
                },
                ClientId: CLIENT_ID
            });

            const response = await this.client.send(command);
            
            this.tokens = response.AuthenticationResult;
            this.username = username;
            
            // Calculate expiration time
            if (this.tokens.ExpiresIn) {
                this.tokens.expires_at = Math.floor(Date.now() / 1000) + this.tokens.ExpiresIn;
            }

            return {
                success: true,
                tokens: this.tokens
            };

        } catch (error) {
            console.error('Login error:', error);
            return {
                success: false,
                error: error.message || 'Error de autenticación'
            };
        }
    }

    async getUserAttributes() {
        if (!this.tokens || !this.tokens.AccessToken) {
            throw new Error("No token disponible. Haz login primero.");
        }

        try {
            const command = new GetUserCommand({
                AccessToken: this.tokens.AccessToken
            });

            const response = await this.client.send(command);
            
            const attributes = {};
            response.UserAttributes.forEach(attr => {
                attributes[attr.Name] = attr.Value;
            });

            return attributes;

        } catch (error) {
            console.error('Get user attributes error:', error);
            throw error;
        }
    }

    async logout() {
        if (this.tokens && this.tokens.AccessToken) {
            try {
                const command = new GlobalSignOutCommand({
                    AccessToken: this.tokens.AccessToken
                });

                await this.client.send(command);
            } catch (error) {
                console.error('Logout error:', error);
            }
        }
        
        this.tokens = null;
        this.username = null;
    }

    getAccessToken() {
        return this.tokens ? this.tokens.AccessToken : null;
    }

    isLoggedIn() {
        return this.tokens && this.tokens.AccessToken && 
               (!this.tokens.expires_at || Date.now() / 1000 < this.tokens.expires_at - 60);
    }

    //TODO: Agregar el refresh del token
}

module.exports = CognitoHandler;